<a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a align="left" href="https://ultralytics.com/yolov5" target="_blank">
<img width="1024", src="https://user-images.githubusercontent.com/26833433/125273437-35b3fc00-e30d-11eb-9079-46f313325424.png"></a>

This is the ** YOLOv5m 🚀 notebook ** by **BST**

# Setup

Install dependencies and check PyTorch and GPU.

In [ ]:
%cd yolov5
%pip install -qr requirements.txt  # install

import torch
import utils
display = utils.notebook_init()  # checks

# 1. COCO train
Download [COCO test2017](https://github.com/ultralytics/yolov5/blob/74b34872fdf41941cddcf243951cdb090fbac17b/data/coco.yaml#L15) dataset (7GB - 40,000 images), to test model accuracy on val set (**5,000 images**).

In [ ]:
### Train YOLOv5m on COCO for
!python train.py --img 640 --batch 128 --epochs 300 --data coco.yaml --cfg models/yolov5m.yaml --cache

wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: (30 second timeout) 
wandb: W&B disabled due to login timeout.
train: weights=, cfg=models/yolov5m.yaml, data=data/coco.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=300, batch_size=128, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=None, image_weights=False, device=2,3, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=runs/train, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
github: skipping check (not a git repository), for updates see https://github.com/ultralytics/yolov5
Python 3.7.0 required by YOLOv5, but Python 3.6.9 is currently installed
YOLOv5 🚀 2022-8-14 Pytho

# 2. export ONNX

In [1]:
#运行注意注释掉yolo.py中detect类中的reshape和permute等后处理操作!
!python export.py --weights weights/best.pt --batch-size 1 --device cpu 

export: data=data/coco128.yaml, weights=['weights/best.pt'], imgsz=[640, 640], batch_size=1, device=cpu, half=False, inplace=False, train=False, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=12, verbose=False, workspace=4, nms=False, agnostic_nms=False, topk_per_class=100, topk_all=100, iou_thres=0.45, conf_thres=0.25, include=['onnx']
YOLOv5 🚀 2022-8-14 Python-3.7.5 torch-1.7.0 CPU

Fusing layers... 
YOLOv5m summary: 290 layers, 21168285 parameters, 0 gradients, 48.1 GFLOPs

PyTorch: starting from weights/best.pt with output shape (1, 255, 80, 80) (40.8 MB)

ONNX: starting export with onnx 1.9.0...
ONNX: export success, saved as weights/best.onnx (80.8 MB)

Export complete (2.48s)
Results saved to /home/xlh/disk2/yolov5/weights
Detect:          python detect.py --weights weights/best.onnx 
Validate:        python val.py --weights weights/best.onnx 
PyTorch Hub:     model = torch.hub.load('ultralytics/yolov5', 'custom', 'weights/best.onnx')
Visualize:   

# 3. Validate
Validate a model's accuracy on [COCO](https://cocodataset.org/#home) val dataset,and do inference on a variety of sources.

##  COCO val
Download [COCO val 2017](https://github.com/ultralytics/yolov5/blob/74b34872fdf41941cddcf243951cdb090fbac17b/data/coco.yaml#L14) dataset (1GB - 5000 images), and test model accuracy.

In [ ]:
# Download COCO val
torch.hub.download_url_to_file('https://ultralytics.com/assets/coco2017val.zip', 'tmp.zip')
!unzip -q tmp.zip -d ../datasets && rm tmp.zip

In [3]:
# Run YOLOv5m(best.pt/best.onnx/quant_model.onnx) on COCO val
!python val.py --weights weights/quant_model.onnx --data coco.yaml --img 640 --iou 0.65 --half

val: data=data/coco.yaml, weights=runs/train/exp/weights/quant_model.onnx, batch_size=1, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=cpu, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs/val, name=exp, exist_ok=False, half=False, dnn=False
Python 3.7.0 required by YOLOv5, but Python 3.6.9 is currently installed
YOLOv5 🚀 2022-8-14 Python-3.6.9 torch-1.9.1+cu111 CPU

Loading runs/train/exp/weights/quant_model.onnx for ONNX Runtime inference...
Python 3.7.0 required by YOLOv5, but Python 3.6.9 is currently installed
2022-09-06 09:29:01.147075417 [W:onnxruntime:, graph.cc:1074 Graph] Initializer 370 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool

2022-09-06 09:29:01.171314862 [W:onnxruntime:, graph.cc:1074 Graph] Initializer Relu_82_n appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 09:29:01.172960282 [W:onnxruntime:, graph.cc:1074 Graph] Initializer Relu_82_range_val appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 09:29:01.172974360 [W:onnxruntime:, graph.cc:1074 Graph] Initializer Relu_82_zero_point appears in grap

Forcing --batch-size 1 square inference (1,3,640,640) for non-PyTorch models
val: Scanning '/home/v100/xie/datasets/coco/val2017.cache' images and labels... 
               Class     Images     Labels          P          R     mAP@.5 mAP@
                 all       5000      36335      0.689      0.539       0.59      0.369
Speed: 0.3ms pre-process, 249.5ms inference, 5.6ms NMS per image at shape (1, 3, 640, 640)

Evaluating pycocotools mAP... saving runs/val/exp2/quant_model_predictions.json...
Python 3.7.0 required by YOLOv5, but Python 3.6.9 is currently installed
loading annotations into memory...
Done (t=0.64s)
creating index...
index created!
Loading and preparing results...
DONE (t=3.85s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=71.45s).
Accumulating evaluation results...
DONE (t=10.99s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.372
 Average Precision  (AP) @[ IoU=0.50      | area=

### best.pt精度
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.429<br>
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.622<br>
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.462<br>
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.251<br>
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.478<br>
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.560<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.340<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.555<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.606<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.416<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.662<br>
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.754<br>
### best.onnx精度
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.429<br>
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.622<br>
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.462<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.251<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.478<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.560<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.340<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.555<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.606<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.416<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.662<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.754<br>
### quant_model.onnx精度
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.372<br>
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.595<br>
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.406<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.215<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.403<br>
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.496<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.506<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.556<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.376<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.600<br>
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.700<br>

## Inference

`detect.py` runs YOLOv5 inference on a variety of sources and save results to `runs/detect`. Example inference sources are:

```shell
python detect.py --source 0  # webcam
                          img.jpg  # image 
                          vid.mp4  # video
                          path/  # directory
                          'path/*.jpg'  # glob
                          'https://youtu.be/Zgi9g1ksQHc'  # YouTube
                          'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```

In [4]:
# Use YOLOv5m(best.pt/best.onnx/quant_model.onnx) to inference
!python detect.py --weights weights/quant_model.onnx --img 640 --conf 0.25 --source data/images/bus.jpg  

detect: weights=['weights/quant_model.onnx'], source=data/images/bus.jpg, data=data/coco.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=cpu, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 2022-8-14 Python-3.7.5 torch-1.7.0 CPU

Loading weights/quant_model.onnx for ONNX Runtime inference...
/home/xlh/disk2/envs/lib/python3.7/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
2022-09-06 15:24:04.265602041 [W:onnxruntime:, graph.cc:1237

2022-09-06 15:24:04.291759538 [W:onnxruntime:, graph.cc:1237 Graph] Initializer model.17.m.1.cv1.conv.bias appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.291765395 [W:onnxruntime:, graph.cc:1237 Graph] Initializer model.17.m.1.cv2.conv.weight appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.291773153 [W:onnxruntime:, graph.cc:1237 Graph] Initializer model.

2022-09-06 15:24:04.339957027 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_38_range_val appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.339970149 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_38_zero_point appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.339980794 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_38_negative_zero_poi

2022-09-06 15:24:04.392765702 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_85_inv_range appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.392770439 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_85_n appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.392775885 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_85_range_val appears in graph

2022-09-06 15:24:04.441964446 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_140_n appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.451070437 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_140_range_val appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2022-09-06 15:24:04.451249170 [W:onnxruntime:, graph.cc:1237 Graph] Initializer Relu_140_zero_point appears in g

image 1/1 /home/xlh/disk2/yolov5/data/images/bus.jpg: 640x640 4 persons, 1 bus, Done. (0.561s)
Speed: 27.6ms pre-process, 561.2ms inference, 22.5ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img align="left" src="runs/detect/exp/bus.jpg" width="500">